### DISCUSSION TITLE
libcef.dll throws an exception when closing the application (code: 0xc0000005)

### DISCUSSION THREAD

**Discussion by christianGen** [cefsharp/CefSharp#4522] [General]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

113.1.40

### Operating System

Windows 10

### Architecture

AnyCPU

### .Net Version

.Net 4.8

### Implementation

WinForms

### Reproduction Steps

Before I try to find a minimal example, please confirm, that this actually is a bug in CefSharp.

### Expected behavior

Don't throw an exception (Exception code: 0xc0000005) when closing.  

### Actual behavior

libcef.dll throws an exception.

### Known Workarounds

Not really a workaround.  But we now avoid dumpfiles using the following code:

```CSharp
        // https://stackoverflow.com/questions/6080605/can-i-use-seterrormode-in-c-sharp-process
        [DllImport("kernel32.dll")]
        static extern ErrorModes  SetErrorMode(ErrorModes uMode);

        [Flags]
        public enum ErrorModes : uint {
            SYSTEM_DEFAULT             = 0x0,
            SEM_FAILCRITICALERRORS     = 0x0001,
            SEM_NOALIGNMENTFAULTEXCEPT = 0x0004,
            SEM_NOGPFAULTERRORBOX      = 0x0002,
            SEM_NOOPENFILEERRORBOX     = 0x8000
        }


            
            // Prevent dumps when crashing in the libcef library.
            SetErrorMode(SetErrorMode(0) | ErrorModes.SEM_NOGPFAULTERRORBOX);
```

### Does this problem also occur in the CEF Sample Application

Not Tested

### Other information

Every time I close my application we get an exception from libcef.dll 113.1.4.0 (earlier versions, for instance 112.3.0.0, do throw that exception as well).

Faulting module name: libcef.dll, version: 112.3.0.0, time stamp: 0x643ef9f4
Exception code: 0xc0000005

I was not able to add symbols to the dump.  The pdb file is definitely the correct version but Visual Studio says "PDB format is not supported".
Before investigation further I would like to confirm that this is a bug in CefSharp.

Can my application be the reason for this exception, or is a bug in CefSharp the only possibility for this exception?
(If it is possible that the application can use libcef incorrectly, please simply close this issue.)

**Comment by amaitland** [comment]
> Before investigation further I would like to confirm that this is a bug in CefSharp.

Crashes in `libcef.dll` are rarely `CefSharp` bugs. It's important to remember that `CefSharp` is a wrapper around the `Chromium Embedded Framework (CEF)`.

Did you check the log file? Test with the `MinimalExample`?

`Cef.Initialize` and `Cef.Shutdown` must be called on the same thread. https://github.com/cefsharp/CefSharp/wiki/General-Usage#initialize-and-shutdown

> The pdb file is definitely the correct version but Visual Studio says "PDB format is not supported".

I remember seeing a report a while ago regarding pdb load issues. https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=19208

If you have the correct `pdb` and correct bitness then you'll need to ask for help on ceforum regarding pdb load issues.

> Can my application be the reason for this exception,

Entirely possible. Impossible to say with the **very** limited information you've provided. It's possible a bug in `CEF` also.

> ### Reproduction Steps
> Before I try to find a minimal example, please confirm, that this actually is a bug in CefSharp.

In future please hold off on creating an issue until you have more information. You can use [Discussions](https://github.com/cefsharp/CefSharp/discussions) to ask for help.

I've not seen any other reports of shutdown issues recently.

**Reply by christianGen** [reply]
Thanks.
Our code is horrible and the chances of us doing something incorrectly are **very** high.

I was just unsure, if `dll`s are "allowed" to throw such exceptions.

Thanks for the initialize / shutdown link.  I will verify that we shutdown correctly.
